external-crates/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -1787,563 +2331,24 @@ impl Frame {
-                        interpreter.operand_stack.push(field_ref)?;
-                    }
-                    Bytecode::ImmBorrowFieldGeneric(fi_idx)
-                    | Bytecode::MutBorrowFieldGeneric(fi_idx) => {
-                        let instr = match instruction {
-                            Bytecode::MutBorrowField(_) => S::MutBorrowFieldGeneric,
-                            _ => S::ImmBorrowFieldGeneric,
-                        };
-                        gas_meter.charge_simple_instr(instr)?;
-
-                        let reference = interpreter.operand_stack.pop_as::<StructRef>()?;
-
-                        let offset = resolver.field_instantiation_offset(*fi_idx);
-                        let field_ref = reference.borrow_field(offset)?;
-                        interpreter.operand_stack.push(field_ref)?;
-                    }
-                    Bytecode::Pack(sd_idx) => {
-                        let field_count = resolver.field_count(*sd_idx);
-                        let struct_type = resolver.get_struct_type(*sd_idx);
-                        self.check_depth_of_type(resolver, &struct_type)?;
-                        gas_meter.charge_pack(
-                            false,
-                            interpreter.operand_stack.last_n(field_count as usize)?,
-                        )?;
-                        let args = interpreter.operand_stack.popn(field_count)?;
-                        interpreter
-                            .operand_stack
-                            .push(Value::struct_(Struct::pack(args)))?;
-                    }
-                    Bytecode::PackGeneric(si_idx) => {
-                        let field_count = resolver.field_instantiation_count(*si_idx);
-                        let ty = resolver.instantiate_generic_type(*si_idx, self.ty_args())?;
-                        self.check_depth_of_type(resolver, &ty)?;
-                        gas_meter.charge_pack(
-                            true,
-                            interpreter.operand_stack.last_n(field_count as usize)?,
-                        )?;
-                        let args = interpreter.operand_stack.popn(field_count)?;
-                        interpreter
-                            .operand_stack
-                            .push(Value::struct_(Struct::pack(args)))?;
-                    }
-                    Bytecode::Unpack(_sd_idx) => {
-                        let struct_ = interpreter.operand_stack.pop_as::<Struct>()?;
-
-                        gas_meter.charge_unpack(false, struct_.field_views())?;
-
-                        for value in struct_.unpack()? {
-                            interpreter.operand_stack.push(value)?;
-                        }
+                match Self::execute_instruction(
+                    &mut self.pc,
+                    &mut self.locals,
+                    &self.ty_args,
+                    &self.function,
```

Comment:
why not pass `&mut self` here? it seems it would be cleaner and it should not be a problem?

Reply:
this prepares things for pre & post hooks which will will rely on components in `self`. Borrow checker will complain if `self` is borrowed as mutable.

Reply:
I am unclear why it would complain if `self` is borrow mutably but not part of it are borrowed mutably (though who knows).
Also I do not think the call to the execution and the pre/post we have should be the same.
But overall not a big deal and we can correct, change at any time if needed